File: src/bin/2.rs
use std::io;
use std::io::BufRead;

use regex::Regex;

fn main() -> io::Result<()> {
    let re: Regex = Regex::new("(\\w+) (\\d+)").unwrap();
    let stdin = io::stdin();
    let mut aim: i64 = 0;
    let mut depth: i64 = 0;
    let mut horiz: i64 = 0;
    for line in stdin.lock().lines() {
        let line = line.unwrap();
        let cap = re.captures(line.trim()).unwrap();
        let arg1: i64 = cap[2].parse().unwrap();
        print!("{}\t{}\t{}", line, &cap[1], arg1);
        match &cap[1] {
            "forward" => { horiz += arg1; depth += aim*arg1 }
            "down" => aim += arg1,
            "up" => aim -= arg1,
            x => {
                println!("arg0 invalid={}", x);
                panic!();
            }
        }
        println!("\t\tdepth={} * horiz={} => {}", depth, horiz, depth * horiz);
    }
    println!("depth={} * horiz={} => {}", depth, horiz, depth * horiz);
    Ok(())
}

File: julia/2.jl
function part1()
    horiz = 0
    depth = 0
    open("2.input") do f
        for line in readlines(f)
            line = split(line)
            dir = line[1]
            num = parse(Int64, line[2])
            if dir == "forward"
                horiz += num
            elseif dir == "down"
                depth += num
            elseif dir == "up"
                depth -= num
            else
                error("bad direction '$dir'")
            end
        end
    end
    println("horiz=$horiz depth=$depth ans=$(horiz*depth)")
end

function part2()
    horiz = 0
    depth = 0
    aim = 0
    open("2.input") do f
        for line in readlines(f)
            line = split(line)
            dir = line[1]
            num = parse(Int64, line[2])
            if dir == "forward"
                horiz += num
                depth += aim * num
            elseif dir == "down"
                aim += num
            elseif dir == "up"
                aim -= num
            else
                error("bad direction '$dir'")
            end
        end
    end
    println("horiz=$horiz depth=$depth aim=$aim ans=$(horiz*depth)")
end

part1()
part2()